### PR TITLE
[Feature] Add support for passing temporal clauses to external MySQL tables

### PR DESCRIPTION
**Pull request by fulghum**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
Fixes # https://github.com/StarRocks/starrocks/issues/10004

## Problem Summary(Required) ：

Temporal queries on system versioned tables are a feature introduced in SQL 2011. Although MySQL does not yet support them, several database management systems in the MySQL-family do (e.g. MariaDB, DoltDB). Being able to query historical data is a powerful feature for analytics use cases. This PR adds the ability to pass temporal clauses in queries to external tables backed by an external MySQL database. https://github.com/StarRocks/starrocks/issues/10004 shows one example of a customer looking for this ability. 

Thank you for considering this change and providing any feedback. 🙏  I have submitted my CLA and have included some concrete usage examples below, too. 

Querying a StarRocks external MySQL table, backed by a DoltDB that supports temporal query syntax:
```sql
mysql> select * from mysql_external_table1;
+------+------------+
| pk   | value      |
+------+------------+
|    1 | one        |
|    2 | twoooo     |
|  100 | one hundy! |
+------+------------+
3 rows in set (0.00 sec)

mysql> select * from mysql_external_table1 as of '0pe6o4rg5fkb8jlippjf4s50aht09u50';
+------+-------+
| pk   | value |
+------+-------+
|    1 | one   |
+------+-------+
1 row in set (0.02 sec)

mysql> select * from mysql_external_table1 as of 'does_not_exist';
ERROR 1064 (HY000): mysql query failed. Err: branch not found: does_not_exist
```

Querying a non-external table correctly returns an error, since temporal queries are not supported there:
```sql
mysql> select * from sr_member for system_time as of timestamp 'not_supported';
ERROR 1064 (HY000): unsupported table type for temporal clauses: OLAP; only external MYSQL tables support temporal clauses
```

## Checklist:

- [X] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 21 / 34 (61.76%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/MysqlScanNode.java|2|8|25.00%|[63, 64, 67, 104, 105, 152]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalMysqlScanOperator.java|1|3|33.33%|[37, 38]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|1|3|33.33%|[250, 251]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalMysqlScanOperator.java|1|3|33.33%|[34, 35]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/MysqlScanImplementationRule.java|1|2|50.00%|[31]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|3|3|100.00%|[]|

**Comment by imay** [comment]
@fulghum 
This PR is so excellent. We are considering supporting time-travel queries.
So your patch supports DoltDB and can be used for our future time-travel query.

I very love this patch. Thank you very much.

**Comment by fulghum** [comment]
Thank you so much @imay! 😁 Awesome to hear you are already thinking about time-travel queries – soo many neat things you can do with that capability! 

StarRocks has been a fun codebase to start to explore. Thank you for taking the time to look through this change. 🙏

**Comment by mofeiatwork** [comment]
Please run the clang-format for changed code.

File: be/src/exec/mysql_scanner.cpp [hunk 1]
```diff
@@ -117,7 +117,7 @@ Status MysqlScanner::query(const std::string& query) {
 Status MysqlScanner::query(const std::string& table, const std::vector<std::string>& fields,
                            const std::vector<std::string>& filters,
                            const std::unordered_map<std::string, std::vector<std::string>>& filters_in,
-                           std::unordered_map<std::string, bool>& filters_null_in_set, int64_t limit) {
+                           std::unordered_map<std::string, bool>& filters_null_in_set, int64_t limit, std::string& temporal_clause) {
```
Comment (mofeiatwork) [inline be/src/exec/mysql_scanner.cpp:120]: Please consider make it `const std::string&` .

**Review by murphyatwork — DISMISSED** [review]

File: be/src/exec/mysql_scanner.cpp (hunk 1 above)
Comment (fulghum) [reply be/src/exec/mysql_scanner.cpp]: Added in latest commit

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 21 / 34 (61.76%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/MysqlScanNode.java|2|8|25.00%|[63, 64, 67, 104, 105, 152]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalMysqlScanOperator.java|1|3|33.33%|[37, 38]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|1|3|33.33%|[250, 251]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalMysqlScanOperator.java|1|3|33.33%|[34, 35]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/MysqlScanImplementationRule.java|1|2|50.00%|[31]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|3|3|100.00%|[]|

**Comment by fulghum** [comment]
Thank you @mofeiatwork! clang_format issues are fixed now.

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4 [hunk 2]
```diff
@@ -210,6 +210,7 @@ NOT: 'NOT';
 NULL: 'NULL';
 NULLS: 'NULLS';
 OBSERVER: 'OBSERVER';
+OF: 'OF';
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4:213]: OF should added to the `nonReserved` too.

Comment (fulghum) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4:213]: added in 6c9b4cf20

**Comment by LittleBeeBee** [comment]
您好，您的邮件已收到，我会尽快给你回复。

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4 [hunk 3]
```diff
@@ -301,6 +302,7 @@ SUBMIT: 'SUBMIT';
 SUM: 'SUM';
 SYNC: 'SYNC';
 SYSTEM: 'SYSTEM';
+SYSTEM_TIME: 'SYSTEM_TIME';
```
[the remark this answers is not in the document]
Comment (fulghum) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4:305]: added in 6c9b4cf20

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 21 / 34 (61.76%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/MysqlScanNode.java|2|8|25.00%|[63, 64, 67, 104, 105, 152]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalMysqlScanOperator.java|1|3|33.33%|[37, 38]|
|:large_blue_circle:|com/starrocks/sql/analyzer/QueryAnalyzer.java|1|3|33.33%|[250, 251]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalMysqlScanOperator.java|1|3|33.33%|[34, 35]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/MysqlScanImplementationRule.java|1|2|50.00%|[31]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/TableRelation.java|3|3|100.00%|[]|

**Review by imay — APPROVED** [review]
Good Job! Thank you very much
LGTM

**Review by chaoyli — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by fulghum** [comment]
Thank you so much for accepting this contribution! 🙏

**Pull request merged**